### PR TITLE
chore: modernise CMake & Python package building

### PR DESCRIPTION
**Pull request by agoose77**
Thanks for a fantastic repository! I created this PR to upgrade the build machinery in-line with modern Python / CMake best-practices. I'm sure it's not perfect, but I think it's an improvement. I probably broke some development scripts that you'd written, so let me know if they're still needed.

A notable change is that the C++ library is moved to a subdirectory. This is mainly because most Python extension builders want to be the main target, rather than a subdirectory.

This PR
- Separates Python module building to be `pip` friendly via `scikit-build-core`
- Moves C++ to `catima` subdirectory (so that root `CMakeLists.txt` can be build with scikit-build-core)
- Moves C++ to a `src` and `include` layout
- Updates CMake to opinionated "Modern CMake"
- Adds CI configuration for `cibuildwheel` on GitHub actions (instead of via appveyor)

### CODE REVIEW

**Comment by agoose77** [comment]
Currently this won't produce a usable SDist for the extension, because it resolves the `catima` module in the parent directory. This will be fixed in future.

**Comment by hrosiak** [comment]
Hi, 
Thanks for the PR. I really welcome any improvement. 
I have to go through all changes and I have to admit the repository was more c++ oriented and there are some projects which use this repository as development dependencty via cmake. So I need to be sure they are still fine after merging. 
The C++ restructuring is fine for me and was planned at some point.  Just it may be good to keep possibility for other project to "git clone"  and add it as dependency. 
Another point: currently I am building python modules for pip via appveyour. I have no experience with github workflows. So does this step works the same/similar way as if does in appveyoue, ie wheels will be created by the git workflow?

File: CMakeLists.txt [hunk 1]
```diff
@@ -9,7 +9,6 @@ option(PYTHON_MODULE "compile the Catima python module(requires numpy and cython
 option(TESTS "build tests" OFF)
 option(EXAMPLES "build examples" OFF)
 option(APPS "build catima applications" ON)
-option(GLOBAL "build with global, sources are required" OFF)
```
Comment (hrosiak) [inline CMakeLists.txt:10]: GLOBAL is optional library to calculate ion charge states, this must be an option for build

**Review by hrosiak — COMMENTED** [review]
GLOBAL is optional library to calculate ion charge states, this must be an option for build

**Comment by hrosiak** [comment]
c++ build does not seem to work according to readme

**Comment by hrosiak** [comment]
to me it seems using external submodules requires manual initialisation or updates which for user will be inconvenient or some of them will not know what to do compared to FetchContent, which makes everything automatically

File: CMakeLists.txt (hunk 1 above)
Comment (agoose77) [reply CMakeLists.txt:10]: Ah OK, I removed this as I thought it did something else!

**Comment by agoose77** [comment]
> So does this step works the same/similar way as if does in appveyoue, ie wheels will be created by the git workflow?

Yes, you can see an example workflow that does this [here](https://github.com/agoose77/catima/actions/runs/4224332100). All of the wheels are in `artifact.zip`

**Comment by agoose77** [comment]
It does for me. It might be that you're in the wrong folder? If this repo is under `/`, then you need to `cd` to `/catima` to run `cmake`.

**Comment by agoose77** [comment]
> to me it seems using external submodules requires manual initialisation or updates which for user will be inconvenient or some of them will not know what to do compared to FetchContent, which makes everything automatically

Modern Git usage / good defaults make submodules pretty easy to use. However, if you prefer, one could also use `FetchContent`:: https://json.nlohmann.me/integration/cmake/#fetchcontent

**Comment by hrosiak** [comment]
> > to me it seems using external submodules requires manual initialisation or updates which for user will be inconvenient or some of them will not know what to do compared to FetchContent, which makes everything automatically
> 
> It does for me. It might be that you're in the wrong folder? If this repo is under `/`, then you need to `cd` to `/catima` to run `cmake`.

Sure i did from catima subdirectory and I had do to git submodule update --init after initial git clone of the catima repository.

**Comment by hrosiak** [comment]
> > to me it seems using external submodules requires manual initialisation or updates which for user will be inconvenient or some of them will not know what to do compared to FetchContent, which makes everything automatically
> 
> Modern Git usage / good defaults make submodules pretty easy to use. However, if you prefer, one could also use `FetchContent`: https://json.nlohmann.me/integration/cmake/#fetchcontent

This is currently used in my main branch, right ?

**Comment by hrosiak** [comment]
Is it possible to combine main CMakeList.txt in a way it work with cibuildwheel or scitkit-build (with python builders i do not have much experience) and at the same time if another C++ project adds catime as submodule, via external module or FetchContent. 
?

**Comment by agoose77** [comment]
> This is currently used in my main branch, right ?

Ah, yes! You do already use this :)

> Is it possible to combine main CMakeList.txt in a way it work with cibuildwheel or scitkit-build (with python builders i do not have much experience) and at the same time if another C++ project adds catime as submodule, via external module or FetchContent.
?

Yes, and you can already do that, but it needs to happen at a deeper directory. I believe `FetchContent` provides such an option (I was looking at this the other day): https://cmake.org/cmake/help/latest/module/FetchContent.html#overriding-where-to-find-cmakelists-txt

Unfortunately the fact that the `scikit-build-core` `CMakeLists.txt` needs to be at the top-level is a restriction that will be removed in `scikit-build-core` eventually, but for now we need to restructure the repo. `scikit-build` probably doesn't have this restriction, but it's not likely to be well maintained for long in my opinion.

**Comment by hrosiak** [comment]
> Yes, and you can already do that, but it needs to happen at a deeper directory. I believe `FetchContent` provides such an option (I was looking at this the other day): https://cmake.org/cmake/help/latest/module/FetchContent.html#overriding-where-to-find-cmakelists-txt

The problem here is I am not maintainer of other projects and can not change FetchContent options in their CmakeList.txt. 
This may possibly break build of those projects. 

At the moment the users of this library, I am aware of, uses C++ part via Cmake and python module only via pip install pycatima, so naturally I would like to solve that C++ part as well.

**Comment by agoose77** [comment]
@hrosiak so, right now the use of `scikit-build-core` is a technical limitation. However, overall it makes building an extension much nicer and future-proof. With that in mind, a simple solution is to move the Python bindings into a new repository, which could then use `FetchContent`. Would that break many existing users?

**Comment by hrosiak** [comment]
> @hrosiak so, right now the use of `scikit-build-core` is a technical limitation. However, overall it makes building an extension much nicer and future-proof. With that in mind, a simple solution is to move the Python bindings into a new repository, which could then use `FetchContent` to load `catima`. Would that break many existing users?

Most of the "python users" either use just module installed from pip repository or via requirements.txt. 
I know only 1 developer who may still use "compiled" python module directly from source, just because he started to use it before it was available via pip.
So this solution, I would say will break very minimum users if any.

**Comment by agoose77** [comment]
Hmm, OK! It's up to you whether you are happy to merge this PR, of course. If you would like to create a repository for the bindings, e.g. `pycatima`, I'd be happy to split this PR. 

It's potentially possible to use `FetchContent` in the same repository, but I think that's *more* hacky than moving the code into a separate repository. If `scikit-build-core` changes, you could always merge the repositories in the future.

**Comment by hrosiak** [comment]
I would go for separate repositories, it seems the most clean solution.

**Comment by agoose77** [comment]
OK! If you create an empty repo, e.g. `pycatima` (so that you're the owner), I'm happy to make a PR.

**Comment by hrosiak** [comment]
https://github.com/hrosiak/pycatima

**Comment by agoose77** [comment]
@hrosiak this _should_ now only contain the C++ sources.

**Comment by hrosiak** [comment]
> @hrosiak ideally the GLOBAL dependency would be a CMake target so that the `CMakeLists.txt` can be very simple. Do you have an example of where that source code can be found? Or is it an internal code of yours?

background is that atm GLOBAL code's license does not allow me to publish it publicly.  So use cases is who has the code can enable it (by default its disabled when you do not specify -DGLOVBAL). 
In the code there are multiple models to calculate ion charge state so by default GLOBAL model is just not available and user can choose one of other models.

**Comment by agoose77** [comment]
Would you be able to add a `CMakeLists.txt` to it so that one can use `add_subdirectory` to include it into the main program?

**Comment by hrosiak** [comment]
> Would you be able to add a `CMakeLists.txt` to it so that one can use `add_subdirectory` to include it into the main program?

yes that is possible

**Comment by hrosiak** [comment]
I still have some problems with make install step

1. bin files if APPS=ON fails to install
2. include files are copied to destination folder with 2 include subdirs, right ? (at least for linux OS):
-- Installing: local/include/include
-- Installing: local/include/include/catima

**Comment by agoose77** [comment]
> bin files if APPS=ON fails to install

I can't reproduce this:
```
$ mkdir build                                                                      13  09:41
$ cmake -S ~/Git/catima -B build -DCMAKE_INSTALL_PREFIX=$PWD/install -DAPPS=ON     14  09:41
-- The CXX compiler identification is GNU 11.3.0
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Check for working CXX compiler: /usr/bin/c++ - skipped
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- Build type Release
-- Build type: Release
-- install prefix: /tmp/tmp.QLy0pjR0D3/install
-- CMAKE_CXX_COMPILER: /usr/bin/c++
-- Using the multi-header code from /tmp/tmp.QLy0pjR0D3/build/_deps/nlohmann_json-src/include/
-- Configuring done
-- Generating done
-- Build files have been written to: /tmp/tmp.QLy0pjR0D3/build
$ cmake --build build                                                              15  09:41
[  6%] Building CXX object CMakeFiles/catima.dir/src/abundance_database.cpp.o
[ 13%] Building CXX object CMakeFiles/catima.dir/src/calculations.cpp.o
[ 20%] Building CXX object CMakeFiles/catima.dir/src/catima.cpp.o
[ 26%] Building CXX object CMakeFiles/catima.dir/src/cwrapper.cpp.o
[ 33%] Building CXX object CMakeFiles/catima.dir/src/integrator.cpp.o
[ 40%] Building CXX object CMakeFiles/catima.dir/src/material_database.cpp.o
[ 46%] Building CXX object CMakeFiles/catima.dir/src/nucdata.cpp.o
[ 53%] Building CXX object CMakeFiles/catima.dir/src/reactions.cpp.o
[ 60%] Building CXX object CMakeFiles/catima.dir/src/spline.cpp.o
[ 66%] Building CXX object CMakeFiles/catima.dir/src/srim.cpp.o
[ 73%] Building CXX object CMakeFiles/catima.dir/src/storage.cpp.o
[ 80%] Building CXX object CMakeFiles/catima.dir/src/structures.cpp.o
[ 86%] Linking CXX shared library libcatima.so
[ 86%] Built target catima
[ 93%] Building CXX object bin/CMakeFiles/catima_calculator.dir/catima_calculator.cpp.o
[100%] Linking CXX executable catima_calculator
[100%] Built target catima_calculator
$ cmake --install build                                                            16  09:41
-- Install configuration: "Release"
-- Installing: /tmp/tmp.QLy0pjR0D3/install/bin/catima_calculator
-- Set runtime path of "/tmp/tmp.QLy0pjR0D3/install/bin/catima_calculator" to ""
-- Installing: /tmp/tmp.QLy0pjR0D3/install/lib/libcatima.so
-- Installing: /tmp/tmp.QLy0pjR0D3/install/include/include
-- Installing: /tmp/tmp.QLy0pjR0D3/install/include/include/catima
-- Installing: /tmp/tmp.QLy0pjR0D3/install/include/include/catima/constants.h
-- Installing: /tmp/tmp.QLy0pjR0D3/install/include/include/catima/structures.h
-- Installing: /tmp/tmp.QLy0pjR0D3/install/include/include/catima/material_database.h
-- Installing: /tmp/tmp.QLy0pjR0D3/install/include/include/catima/reactions.h
-- Installing: /tmp/tmp.QLy0pjR0D3/install/include/include/catima/catima.h
-- Installing: /tmp/tmp.QLy0pjR0D3/install/include/include/catima/data_atima.h
-- Installing: /tmp/tmp.QLy0pjR0D3/install/include/include/catima/glq_integrator.h
-- Installing: /tmp/tmp.QLy0pjR0D3/install/include/include/catima/config.h
-- Installing: /tmp/tmp.QLy0pjR0D3/install/include/include/catima/srim.h
-- Installing: /tmp/tmp.QLy0pjR0D3/install/include/include/catima/generated_LS_coeff.h
-- Installing: /tmp/tmp.QLy0pjR0D3/install/include/include/catima/calculations.h
-- Installing: /tmp/tmp.QLy0pjR0D3/install/include/include/catima/abundance_database.h
-- Installing: /tmp/tmp.QLy0pjR0D3/install/include/include/catima/gkq_integrator.h
-- Installing: /tmp/tmp.QLy0pjR0D3/install/include/include/catima/cwrapper.h
-- Installing: /tmp/tmp.QLy0pjR0D3/install/include/include/catima/storage.h
-- Installing: /tmp/tmp.QLy0pjR0D3/install/include/include/catima/convert.h
-- Installing: /tmp/tmp.QLy0pjR0D3/install/include/include/catima/nucdata.h
-- Installing: /tmp/tmp.QLy0pjR0D3/install/include/include/catima/integrator.h
-- Installing: /tmp/tmp.QLy0pjR0D3/install/include/include/catima/data_ionisation_potential.h
-- Installing: /tmp/tmp.QLy0pjR0D3/install/include/include/catima/spline.h
```

But as you can see, I also reproduce the bad header installation. That's a regression since we moved the python module. Let me take a look.

**Pull request merged**